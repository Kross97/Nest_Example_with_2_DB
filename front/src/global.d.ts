
// обязательная установка sass пакета (npm i sass --save-dev)

declare module '*.module.scss' {
  const css: Record<string, string>;
  export default css;
}

declare module '*.scss' {
  const css: Record<string, string>;
  export default css;
}