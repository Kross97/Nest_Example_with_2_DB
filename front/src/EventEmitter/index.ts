interface IHandler {
  once: boolean;
  handler: () => void;
}

export const EventsDictionary = {
  unAuthorized: "unAuthorized",
} as const;

type TEvents = keyof typeof EventsDictionary;

class EventEmitter {
  private eventsHandlersDispatcher: Record<string, IHandler[]> = {};

  add(event: TEvents, handler: () => void, once?: boolean) {
    if (event in this.eventsHandlersDispatcher) {
      this.eventsHandlersDispatcher[event].push({ handler, once: !!once });
    } else {
      this.eventsHandlersDispatcher[event] = [{ handler, once: !!once }];
    }
  }

  remove(event: TEvents, handler: () => void) {
    if (event in this.eventsHandlersDispatcher) {
      this.eventsHandlersDispatcher[event] = this.eventsHandlersDispatcher[event].filter(
        (handlerObj) => handlerObj.handler !== handler
      );
    }
  }

  emit(event: TEvents) {
    if (event in this.eventsHandlersDispatcher) {
      this.eventsHandlersDispatcher[event] = this.eventsHandlersDispatcher[event].filter(
        (handlerObj) => {
          handlerObj.handler();
          return !handlerObj.once;
        }
      );
    }
  }
}

export const EventEmitterAgent = new EventEmitter();
