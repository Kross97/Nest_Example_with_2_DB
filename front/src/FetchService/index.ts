import {  getTokenJwt } from '../provider/AllProvider';
import { errorHandler } from './utils/errorHandler';

interface IUrlParams {
  url: string;
  headers?: Record<string, string>
  body?: any;
  requestType?: 'none' | 'json';
  responseType?: 'json' | 'blob' | 'text' | 'arrayBuffer';
  method: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';
}

const dispatcherHeaders = {
  'json': { 'content-type': 'application/json; charset=utf-8' },
  'none': {}
};

const dispatcherBody = {
  'json': (body: any) => JSON.stringify(body),
  'none': (body: any) => body,
}

const initAuthHeader = () => ({ 'Authorization': `Bearer ${getTokenJwt()}`})

class FetchService {

  private backUrl: string = process.env.BACK_HOST || 'http://localhost:3001';

  async request<T = any>({ url, headers = {}, body = '', method, requestType = 'none', responseType = 'json' }: IUrlParams) {
    const urlHost = (this.backUrl + (url.startsWith('/') ? url : `/${url}`));
    const currentHeaders = {...headers, ...dispatcherHeaders[requestType], ...initAuthHeader()};
    const currentUrl = method === 'GET' ? `${urlHost}?${new URLSearchParams(body).toString()}` : urlHost;

    const response = await fetch(currentUrl, {
      headers: currentHeaders,
      method,
      body: method === 'GET' ? null : dispatcherBody[requestType](body),
      credentials: 'include'
    });

    if(response.ok) {
      const contentLength = Number(response.headers.get('content-length') || 0);
      if (contentLength) {
        return response[responseType]();
      }
    } else {
      errorHandler(response);
    }
  }

  async postRequest<T = any>(data: Omit<IUrlParams, 'method'>) {
    return this.request<T>({ ...data, method: 'POST' });
  }

  async getRequest<T = any>(data: Omit<IUrlParams, 'method'>) {
    return this.request<T>({ ...data, method: 'GET' });
  }

  async putRequest<T = any>(data: Omit<IUrlParams, 'method'>) {
    return this.request<T>({ ...data, method: 'PUT' });
  }
};

export const FetchAgent = new FetchService();