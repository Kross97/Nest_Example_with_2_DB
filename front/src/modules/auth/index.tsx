import cn from './Auth.module.scss';
import React, { ChangeEvent,  useState } from 'react';
import { FetchAgent } from '../../FetchService';
import { useAllContext } from '../../provider/AllProvider';


export const AuthBlock = () => {
  const [data, setUser] = useState<Record<string, string>>({});

  const { setToken, setCurrentPortHandler } = useAllContext();

  const change = (e: ChangeEvent<HTMLInputElement>) => {
    const nameField = e.target.name;
    const value = e.target.value;
    setUser((prev) => ({ ...prev, [nameField]: value }));
  };

  const signIn = async () => {
     const tokenData = await FetchAgent.postRequest<{ token: string; currentPort: number}>({ url: 'auth/signIn', requestType: 'json', body: data });
     console.log('tokenData =>', tokenData)
     setToken(tokenData.token);
     setCurrentPortHandler(tokenData.currentPort);
  };

  return <div className={cn.authBlock}>
    <h3>Авторизация</h3>
    <span>Логин</span>
    <input onChange={change} type={'text'} name={'login'} value={data.login}/>
    <span>Пароль</span>
    <input onChange={change} type={'password'} name={'password'} value={data.password}/>
    <button onClick={signIn}>Войти</button>
  </div>;
};