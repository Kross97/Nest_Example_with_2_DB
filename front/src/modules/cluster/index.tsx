import React, { useEffect, useState } from "react";
import cn from "./Cluster.module.scss";
import { FetchAgent } from "../../FetchService";
import { IPortsData, useAllContext } from "../../provider/AllProvider";

export function ClusterBlock() {
  const [currentPorts, setPorts] = useState<number[]>([]);

  const { setCurrentPortHandler, currentPort } = useAllContext();

  const getCurrentPorts = () => {
    FetchAgent.getRequest({ url: "/childProcess_cluster/cluster_ports" })
      .then((results: IPortsData) => {
        setPorts(results.availablePorts);
      })
      .catch((err) => {
        console.log("Ошибка получения портов", Object.entries(err));
      });
  };

  useEffect(() => {
    getCurrentPorts();
  }, []);

  const enableClusterHandle = () =>
    FetchAgent.getRequest({ url: "/childProcess_cluster/cluster_enable" }).then(() => {
      getCurrentPorts();
    });

  const disableClusterHandler = () => {
    FetchAgent.getRequest({ url: "/childProcess_cluster/cluster_exit" }).then(() => {
      getCurrentPorts();
      setCurrentPortHandler();
    });
  };

  const setNewPortServer = () => {
    void setCurrentPortHandler();
  };

  return (
    <div className={cn.clusterBlock}>
      <h2>Работа с кластеризацией сервера (Cluster модуль node)</h2>
      <h3>{`Состояние кластеризации: ${currentPorts.length === 1 ? "Отключена" : "Включена"}`}</h3>
      <h3>{`Текущий порт: ${currentPort}`}</h3>
      <div>
        <span>Доступные порты:</span>
        <span>
          {currentPorts.map((currentPort) => (
            <React.Fragment key={currentPort}>
              <span style={{ cursor: "pointer", color: "blue" }} onClick={() => setCurrentPortHandler(currentPort)}>
                {currentPort}
              </span>
              <span>{` , `}</span>
            </React.Fragment>
          ))}
        </span>
      </div>
      <div>
        <button disabled={currentPorts.length === 1} onClick={setNewPortServer}>
          Выбрать новый порт сервера
        </button>
        <button disabled={currentPorts.length !== 1} onClick={enableClusterHandle}>
          Включить кластеризацию
        </button>
        <button disabled={currentPorts.length === 1} onClick={disableClusterHandler}>
          Отключить кластеризацию
        </button>
      </div>
    </div>
  );
}
