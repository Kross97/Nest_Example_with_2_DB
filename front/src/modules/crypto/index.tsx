import cn from './Crypto.module.scss';
import React, { useEffect, useState } from 'react';
import { FetchAgent } from '../../FetchService';


export const CryptoBlock = () => {
  const [search, setSearch] = useState('');
  const [cipherData, setCipherData] = useState('');
  const [deCipherData, setDeCipherData] = useState('');


  const cipherHandler = async () => {
   const res = await FetchAgent.postRequest({ url: 'crypto/cipher1', body: search})
   setCipherData(res);
  };

  const deCipherHandler = async () => {
    const res = await FetchAgent.postRequest({ url: 'crypto/decipher1', body: cipherData })
    setDeCipherData(res);
  };

  return  <div className={cn.cryptoBlock}>
    <h3>Модуль для работы с шифрованием (Crypto node.js)</h3>
    <h4>Шифрование и де-шифрование данных</h4>
    <label className={cn.column}>
      Данные для шифрования
      <input type={'text'} value={search} onChange={(e) => setSearch(e.target.value)} />
    </label>
    <div className={cn.column}>
      <span>Данные после шифрования</span>
      <span>{cipherData}</span>
    </div>
    <div className={cn.column}>
      <span>Данные после де-шифрования</span>
      <span>{deCipherData}</span>
    </div>
    <button onClick={cipherHandler}>шифровать</button>
    <button onClick={deCipherHandler}>де-шифровать</button>
  </div>
};