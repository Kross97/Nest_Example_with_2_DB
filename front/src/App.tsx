import React from 'react';
import './App.css';
import { PhotoBlock } from './modules/photos';
import { UserBlock } from './modules/user';
import { AuthBlock } from './modules/auth';
import {  useAllContext } from './provider/AllProvider';
import { CryptoBlock } from './modules/crypto';

function App() {
  const { token, setToken } = useAllContext();

  const exitHandler = () => {
    setToken(null);
  };

  return (
    <>
      {token && <button onClick={exitHandler} style={{ position: 'fixed', top: '50px', right: '50px'}}>exit</button>}
      <div style={{ display: 'flex'}}>
        {token && <PhotoBlock />}
        {token && <UserBlock />}
        {token && <CryptoBlock />}
        {!token && <AuthBlock />}
      </div>
    </>
  );
}

export default App;
