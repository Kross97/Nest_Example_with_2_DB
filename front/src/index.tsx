import React from 'react';
import ReactDOM from 'react-dom/client';
import './index.css';
import { FullApp } from './App';
import { AllProvider } from './provider/AllProvider';
import './FetchService/index';

const root = ReactDOM.createRoot(
  document.getElementById('root') as HTMLElement
);
root.render(
  <React.StrictMode>
    <AllProvider>
      <FullApp />
    </AllProvider>
  </React.StrictMode>
);
