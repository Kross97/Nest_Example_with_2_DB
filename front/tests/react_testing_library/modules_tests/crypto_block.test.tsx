import React from 'react';
import { CryptoBlock } from '../../../src/modules/crypto';
import { render, screen, cleanup } from 'test-utils';
import { FetchAgent } from '../../../src/FetchService';
import { userEvent } from '@testing-library/user-event';

beforeAll(() => {
  // console.log('crypto', FetchAgent.postRequest);
});


afterEach(() => {
  // Демонтирует компонент из контейнера и уничтожает контейнер. (для того чтобы каждый тест начинался с нуля)
  cleanup();
});

const dataForCipher = 'Данные для шифрования';
const dataCipher = 'данные после шифрования (cipher)'

const dataForDecipher = 'Данные для де-шифрования';
const dataDeCipher = 'Данные после де-шифрования (decipher)';

const signData = 'Данные для подписи';

describe('Тестирование модуля Crypto с помощью user-events', () => {
  test.skip('Тестирование работы Шифрования (Cipher)', async () => {
    render(<CryptoBlock />)

    const inputCipher = screen.getByLabelText(dataForCipher, {
      exact: false,
    });

    const mockPost = jest.spyOn(FetchAgent, 'postRequest').mockImplementation(() => Promise.resolve(dataCipher));

    await userEvent.type(inputCipher, dataForCipher);
    await userEvent.click(screen.getByTestId('cipher-button'))

    expect(mockPost).toHaveBeenCalled();
    expect(mockPost.mock.calls[0][0].body).toBe(dataForCipher);
    expect(screen.getByTestId('cipher-field')).toHaveTextContent(dataCipher);
  });

  test.skip('Тестирование Де-шифрования (Decipher)', async () => {
    render(<CryptoBlock />)

    const input = screen.getByLabelText(dataForDecipher, {
      exact: false,
    })

    const mockPost = jest.spyOn(FetchAgent, 'postRequest').mockImplementation(() => Promise.resolve(dataDeCipher));

     await userEvent.type(input, dataForDecipher);
     await userEvent.click(screen.getByTestId('decipher-button'));


    expect(mockPost).toHaveBeenCalled();
    expect(mockPost.mock.calls[0][0].body).toBe(dataForDecipher);
    expect(screen.getByTestId('decipher-field')).toHaveTextContent(dataDeCipher)
  });

  test('Тестирование Снапшотом', async () => {
      const { container } = render(<CryptoBlock />);

     const inputSign = screen.getByLabelText(signData, {
       exact: false,
     });

     const inputCipher = screen.getByLabelText(dataForCipher, {
       exact: false,
     })

    const inputDecipher = screen.getByLabelText(dataForDecipher, {
      exact: false,
    })

    await userEvent.type(inputSign, signData);
    await userEvent.type(inputCipher, dataForCipher);
    await userEvent.type(inputDecipher, dataForDecipher);

    expect(screen.getByTestId('cipher-field')).toHaveTextContent(dataForDecipher);
    expect(container).toMatchSnapshot();
  });
});