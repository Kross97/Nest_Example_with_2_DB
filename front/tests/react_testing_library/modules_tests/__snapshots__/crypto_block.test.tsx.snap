// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тестирование модуля Crypto с помощью user-events Тестирование Снапшотом 1`] = `
<div>
  <div
    style="display: flex; column-gap: 10px; white-space: wrap; word-break: break-all;"
  >
    <div
      class="cryptoBlock"
    >
      <h3>
        Модуль для работы с шифрованием (Crypto node.js)
      </h3>
      <h4>
        Шифрование и де-шифрование данных
      </h4>
      <div
        class="blockInfo"
      >
        <span>
          <b>
            1. "+$+"
          </b>
           - в шифрованном тексте является разделителем между вектором инициализации 
          <b>
            (iv)
          </b>
           и шифруемыми данными
        </span>
        <span>
          <b>
            2. iv (вектор инициализации)
          </b>
           - участвует в шифровании данных, при каждом новом шифровании 
          <b>
            (Cipher)
          </b>
           возвращаемый шифр будет отличаться т.к создается новый экземпляр iv (вектора инициализации)
        </span>
        <span>
          <b>
            3. iv берется из шифра по разделителю
          </b>
           - при каждой де-шифрации, что гарантирует корректную работу де-шифрации 
          <b>
            (Decipher)
          </b>
        </span>
      </div>
      <label
        class="column"
      >
        Данные для шифрования 
        <b>
          (класс Cipher node)
        </b>
        <input
          type="text"
          value="Данные для шифрования"
        />
      </label>
      <div
        class="column"
      >
        <span>
          Данные после шифрования 
          <b>
            (класс Decipher node)
          </b>
        </span>
        <span
          data-testid="cipher-field"
        >
          Данные для де-шифрования
        </span>
      </div>
      <label
        class="column"
      >
        Данные для де-шифрования (hex - строка) 
        <b>
          (класс Decipher node)
        </b>
        <input
          type="text"
          value="Данные для де-шифрования"
        />
      </label>
      <div
        class="column"
      >
        <span>
          Данные после де-шифрования
        </span>
        <span
          data-testid="decipher-field"
        />
      </div>
      <button
        data-testid="cipher-button"
      >
        шифровать
      </button>
      <button
        data-testid="decipher-button"
      >
        де-шифровать
      </button>
    </div>
    <div
      class="cryptoSignBlock"
    >
      <h3>
        Модуль для работы с подписями (Sign, Verify node.js)
      </h3>
      <h4>
        Подпись и проверка данных
      </h4>
      <label
        class="column"
      >
        Данные для подписи 
        <b>
          (класс Sign, Verify node)
        </b>
        <input
          type="text"
          value="Данные для подписи"
        />
      </label>
      <button>
        тест подписи
      </button>
    </div>
  </div>
</div>
`;
